src/meepgeom.cpp

```diff
@@ -2664,45 +2666,60 @@ std::complex<double> get_material_gradient(
       v.set_direction_max(d, r.in_direction(d) + 0.5*gv.inva*sd);
     }
     double row_1[3], row_2[3], dA_du[3];
-    geps->u_p = -du;
+    double orig = u[idx];
+    u[idx] -= du;
     geps->eff_chi1inv_row(adjoint_c, row_1, v, geps->tol, geps->maxeval);
-    geps->u_p = du;
+    u[idx] += 2*du;
     geps->eff_chi1inv_row(adjoint_c, row_2, v, geps->tol, geps->maxeval);
-    geps->u_p=0;
+    u[idx] = orig;
 
     for (int i=0;i<3;i++) dA_du[i] = (row_1[i] - row_2[i])/(2*du);
     return dA_du[dir_idx] * fields_f;
   
   }else{
+    double orig = u[idx];
     std::complex<double> row_1[3], row_2[3], dA_du[3];
-    geps->u_p = -du;
+    u[idx] -= du;
     eff_chi1inv_row_disp(adjoint_c,row_1,r,freq,geps);
-    geps->u_p = du;
+    u[idx] += 2*du;
     eff_chi1inv_row_disp(adjoint_c,row_2,r,freq,geps);
-    geps->u_p=0;
+    u[idx] = orig;
 
     for (int i=0;i<3;i++) dA_du[i] = (row_1[i] - row_2[i])/(2*du);
     return dA_du[dir_idx] * fields_f;
   }  
 }
 
-/* add the weights from linear_interpolate (see the linear_interpolate
-   function in fields.cpp) to data ... this has to be changed if
-   linear_interpolate is changed!! ...also multiply by scaleby
-   etc. for different gradient types */
-
+/* A brute force approach to calculating Aᵤ using finite differences.
+Past versions of the code only calculated dA/dε using a finite
+difference, and then multiplied by the analytic vJp (dε/du).
+With the addition of subpixel smoothing, however, the vJp became
+much more complicated and it is easier to calculate the entire gradient
+using finite differences (at the cost of slightly less gradients due to 
```

Comment:
"slightly less accurate gradients" you mean?

Reply:
Fixed